### PR TITLE
`function/direxists`: Add function that checks if a directory exists

### PR DESCRIPTION
**Pull request by austinvalle**
Ref: https://github.com/hashicorp/terraform/issues/33394, https://github.com/hashicorp/terraform/issues/25316, https://github.com/hashicorp/terraform/pull/29996

This PR introduces a new provide-defined function for checking the existence of a directory on the filesystem running Terraform. Similar to [`fileexists`](https://developer.hashicorp.com/terraform/language/functions/fileexists).

```tf
terraform {
  required_providers {
    local = {
      source = "hashicorp/local"
    }
  }
  required_version = ">= 1.8.0"
}

output "example_output" {
  value = provider::local::direxists("${path.module}/example-directory")
}
```

```bash
 $ tree          
.
├── example-directory
└── function.tf

2 directories, 1 file

 $ terraform plan

Changes to Outputs:
  + example_output = true

You can apply this plan to save these new output values to the Terraform state, without changing any real infrastructure.

```

### Notes
- I matched the implementation to [`fileexists`](https://developer.hashicorp.com/terraform/language/functions/fileexists) for consistency, but if we prefer deviating we could make some changes
  - Biggest sticking point seems to be returning an error rather than a `false` if "something" exists at the path but is not a directory. I'm not sure returning `false` in that scenario is better, so I kept with consistency.
  - If we do keep the error, we should consider mimicking the terraform core logic of pretty printing the type of file it actually is in the error message: https://github.com/hashicorp/terraform/blob/8da1c006f9d7f381e644a5f0e2c25094ea29571d/internal/lang/funcs/filesystem.go#L234-L257
- I named the files `function_` to follow the convention of the repository. I can also refactor all the files in this provider to match the other accepted pattern of:
  - `*_function.go`
  - `*_resource.go`
  - `*_data_source.go`

### CODE REVIEW

File: internal/provider/function_direxists.go
Comment (bflad) [inline internal/provider/function_direxists.go:69]: Maybe I'm off-base, but I think a difference is that sockets, named pipes, etc. _appear_ to be files (e.g. they show up in `ls` output similar to regular files) unless inspected deeper (e.g. `ls -l`). I think its harder for directories to be confused in the same manner since many tools have stronger distinctions between files and directories in their output. We can always add this later, if desired, since it would only potentially enhance the error messaging rather than changing the function behavior.

File: internal/provider/function_direxists.go
Comment (bflad) [inline internal/provider/function_direxists.go:52]: One thing that `fileexists()` does that is not done here is automatically expand `~` into the calling user's home directory. I'm not sure we necessarily need to do that in this function, but if that is something we want for consistency, then it is probably best to do that upfront to not introduce a potentially breaking change later.

File: internal/provider/function_direxists_test.go
Comment (bflad) [inline internal/provider/function_direxists_test.go:19]: Submitted: https://github.com/hashicorp/terraform-plugin-testing/pull/278

File: internal/provider/testdata/TestLocalFileDataSource/local_file
Comment (bflad) [inline internal/provider/testdata/TestLocalFileDataSource/local_file:1]: Nit: Should we create this as a regular file and introduce a separate test for a symlink file? Might make future troubleshooting easier should symlink file behavior change.

**Review by bflad — COMMENTED** [review]
Nice! Please reach out with any questions.

File: internal/provider/function_direxists_test.go
Comment (bflad) [reply internal/provider/function_direxists_test.go:19]: Oh! And you can/should use [`tfversion.SkipBelow()`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-testing/tfversion#SkipBelow) so CI can still run against all earlier Terraform versions.

Comment (austinvalle) [reply internal/provider/function_direxists_test.go]: Thanks for submitting that. Also whoopsie 😆, SkipBelow is indeed what I wanted lol. Fixed ⚡

File: internal/provider/testdata/TestLocalFileDataSource/local_file
Comment (austinvalle) [reply internal/provider/testdata/TestLocalFileDataSource/local_file:1]: Added a file extension to this existing file test + added a symlink test

File: internal/provider/function_direxists.go
Comment (austinvalle) [reply internal/provider/function_direxists.go]: Makes sense to me, I'm cool with leaving this as-is. Removed the comment ⚡

File: internal/provider/function_direxists.go
Comment (austinvalle) [reply internal/provider/function_direxists.go:65]: Does d993daa match what you were suggesting? I saw that `fileexists` uses a set base directory and an external `go-homedir` library to expand that path. I believe this commit should behave in the same way

Comment (bflad) [reply internal/provider/function_direxists.go:65]: Hmm. I don't think stdlib does anything with regards to `~` expansion unless that's changed in newer versions of Go? In general though, the absolute path translation and cleaning is practitioner a benefit here as it should help with troubleshooting whatever relative/funky paths given actually mean to the provider logic if the file system calls generate an error.

**Comment by bflad** [comment]
terraform-plugin-docs is now updated on main 👍

File: examples/functions/direxists/function.tf
Comment (austinvalle) [inline examples/functions/direxists/function.tf:1]: Taking suggestions on a better example here 😆, I know the `./example-directory` won't get pulled into the registry....

I also considered something like:
```
output "example_output" {
  # Returns `true` on linux/mac, returns `false` on windows
  value = provider :: local :: direxists("/usr/local/bin")
}
```

File: examples/functions/direxists/function.tf
Comment (austinvalle) [inline examples/functions/direxists/function.tf:15]: Same formatting note as the time provider, will be fixed before merge by the beta release of Terraform. Good thing cause of the `local` syntax highlighting lol

File: .github/workflows/test.yml
Comment (austinvalle) [inline .github/workflows/test.yml:86]: Updated to:

`["0.12.*","0.13.*","0.14.*","0.15.*","1.0.*","1.1.*","1.2.*","1.3.*","1.4.*","1.5.*","1.6.*", "1.7.*", "1.8.0-alpha20240228"]`

Will revert in a future PR 😃

File: examples/functions/direxists/function.tf
Comment (austinvalle) [reply examples/functions/direxists/function.tf]: Fixed with introduction of beta release: 9c98a52dcaf6051daafd403f055b13ece5efa603

File: .github/workflows/test.yml
Comment (austinvalle) [reply .github/workflows/test.yml]: Successful tests with new `1.8.0-beta1` in `TF_VERSIONS_PROTOCOL_V5`

9c98a52dcaf6051daafd403f055b13ece5efa603 ⚡ 

https://github.com/hashicorp/terraform-provider-local/actions/runs/8177950751/job/22360813566?pr=285

File: examples/functions/direxists/function.tf
Comment (bendbennett) [reply examples/functions/direxists/function.tf:1]: The example you have LGTM, but could include the `/usr/local/bin` example too perhaps?

File: internal/provider/data_source_local_file_test.go
Comment (bendbennett) [inline internal/provider/data_source_local_file_test.go:22]: ❤️

File: internal/provider/function_direxists.go
Comment (bendbennett) [inline internal/provider/function_direxists.go:32]: Is it worth mentioning/clarifying the behaviour with symlink'd dirs, I assume they're considered the same as regular dirs? What would constitute a FIFO in this context?

**Review by bendbennett — APPROVED** [review]
Nice! 

LGTM 🚀

File: internal/provider/function_direxists.go
Comment (bflad) [reply internal/provider/function_direxists.go:65]: Option 3 seems reasonable to me in this situation. I might encourage some upfront form of that `pathexpand()` mentioning documentation around this since it will differ from the native `fileexists()` behavior that practitioners might already be using/familiar. 👍

Comment (austinvalle) [reply internal/provider/function_direxists.go:67]: Cool, I made some explicit comment and added an example in 782da26543ee4b0c3dc89c66e563437037b40f28 ⚡

File: examples/functions/direxists/function.tf
Comment (austinvalle) [reply examples/functions/direxists/function.tf:1]: I also have now included a homedir example in 782da26543ee4b0c3dc89c66e563437037b40f28

I think that should cover our bases now 👍🏻

File: internal/provider/function_direxists.go
Comment (austinvalle) [reply internal/provider/function_direxists.go]: I originally copied this over from `fileexists`, but I think it makes it more confusing because symlink could be considered a special mode lol.

What do you think of 275e9211d28e48859f76278b15801c59b9d5498e ?

**Review by bflad — APPROVED** [review]
Looks good to me 🚀

**Pull request merged**
